### PR TITLE
chore: no error swallowing

### LINKED ISSUE #39: Review error catching
We currently don't catch errors correctly, and smother some caught issues. 

Some examples of this:
- Any non-HTTP errors are swallowed https://github.com/web3-storage/w3name/blob/8c9ea59ebe2ca99826d4da620e6e3cb5df2b3814/packages/api/src/index.ts#L36-L41
- DO Alarm errors are ignored https://github.com/web3-storage/w3name/blob/8c9ea59ebe2ca99826d4da620e6e3cb5df2b3814/packages/api/src/name.ts#L136-L138
- Some of our error messaging is too generic - we could perhaps try and return something more helpfull! https://github.com/web3-storage/w3name/blob/8c9ea59ebe2ca99826d4da620e6e3cb5df2b3814/packages/api/src/name.ts#L141

We should review ALL instances of error catching and ensure we align on how errors are returned to the user.

### PR DESCRIPTION
**Pull request by joshJarr**
Just going through all of our try and catch statements to ensure we're not swallowing any more errors!

Fixes #39.

### CODE REVIEW

File: packages/api/src/name.ts [hunk 1]
```diff
@@ -95,6 +95,7 @@ export async function namePost (request: Request, env: Env, ctx: ExecutionContex
     if (error instanceof Error) {
       throw new HTTPError(`invalid ipns entry: ${error.message}`, 400)
     }
+    throw new HTTPError(`invalid ipns entry: ${error}`, 400)
```
Comment (francois-potato) [inline packages/api/src/name.ts:98]: Just for consistency with other errors maybe we should capitalize Invalid?

File: packages/api/src/index.ts [hunk 2]
```diff
@@ -36,9 +36,24 @@ export default {
...
+        if (typeof error === 'string') {
+          message = error
+        } else if (error instanceof Object) {
+          message = error.message || JSON.stringify(error)
+        } else {
+          message = `Uncaught error: ${error.toString()}`
+        }
+
+        const httpError = new HTTPError(message, status);
+        return jsonResponse(JSON.stringify({ message: httpError}), httpError.status)
```
Comment (francois-potato) [inline packages/api/src/index.ts:54]: Is there any tests we could add to reach this code?

Comment (joshJarr) [reply packages/api/src/index.ts:54]: I believe this error catch/response is tested with a few error tests here: https://github.com/web3-storage/w3name/blob/chore/no-error-swallowing/packages/api/test/name.spec.ts#L59

File: packages/api/src/name.ts (hunk 1 above)
Comment (joshJarr) [reply packages/api/src/name.ts:98]: Thats a good point! - resolved

File: packages/api/src/name.ts [hunk 3]
```diff
@@ -26,7 +26,7 @@ function parseAndValidateCID (key: string): CID {
   try {
     cid = CID.parse(key, base36)
   } catch (err) {
-    throw new HTTPError('invalid key', 400)
+    throw new HTTPError('invalid key: cannot parse CID.', 400)
```
[the remark this answers is not in the document]
Comment (joshJarr) [reply packages/api/src/name.ts:29]: Yep! resolved

File: packages/api/src/name.ts [hunk 4]
```diff
@@ -69,7 +69,10 @@ export async function nameGet (request: Request, env: Env): Promise<Response> {
     if (error instanceof HTTPError) {
       throw (error)
     } else {
-      throw new HTTPError('please try again', 500)
+      const errorString = error instanceof Object
+        ? JSON.stringify(error)
+        : 'unknown error'
+      throw new HTTPError(`Unable to retrieve name. ${errorString}`, 500)
```
Comment (adamalton) [inline packages/api/src/name.ts:75]: Is the more likely scenario here that `error` is an instance of `Error`, and therefore we should use `error.message`? Calling `JSON.stringify(error)` when `error` is an `Error` just gives `'{}'`.

File: packages/client/src/index.ts [hunk 5]
```diff
@@ -249,6 +249,8 @@ async function maybeHandleError (resPromise: Promise<globalThis.Response>): Prom
   const err = new Error(`unexpected status: ${res.status}`)
   try {
     Object.assign(err, await res.json())
-  } catch {}
+  } catch {
+    throw err
+  }
   throw err
```
Comment (adamalton) [inline packages/client/src/index.ts:255]: The `catch` scenario will still do `throw err` anyway, right? So the behaviour here doesn't actually change.

File: packages/api/src/index.ts [hunk 6]
```diff
@@ -41,9 +41,24 @@ export default {
...
+          message = JSON.stringify(error)
+        } else {
+          message = 'Uncaught error'
+        }
+
+        const httpError = new HTTPError(message, 500)
+        return addCorsHeaders(
+          request,
+          jsonResponse(JSON.stringify({ message: httpError.message }), httpError.status)
+        )
```
[the remark this answers is not in the document]
Comment (joshJarr) [reply packages/api/src/index.ts:59]: oh yeah, totally needlessly casting this to a HTTPError! Will just use message and 500 as the status. Thanks!

File: packages/client/src/index.ts (hunk 5 above)
Comment (joshJarr) [reply packages/client/src/index.ts:255]: Yeah I could possibly just wrap this all in a `finally`

Comment (adamalton) [reply packages/client/src/index.ts]: Ah, an excellent use of the old `finally` statement.

**Review by adamalton — APPROVED** [review]
LGTM

**Review by adamalton — APPROVED** [review]
LGTM

**Pull request merged**